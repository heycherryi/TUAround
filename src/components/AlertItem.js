import React from 'react';
import PropTypes from 'prop-types';
import { renderUnixTimestamp } from '../util';
import dayjs from 'dayjs';
import './AlertItem.scss';
import { Link } from 'react-router-dom';
import StopCode from './StopCode.js';

function AlertItem({ alert, route }) {
  const alertStyle = {
    borderColor: (route.route_color) ? `#${route.route_color}` : '#eee',
    backgroundColor: (route.route_color) ? `#${route.route_color}` : '#eee',
    color: (route.route_text_color) ? `#${route.route_text_color}` : '#000',
  };

  const htmlCleanupRegex = /(<([^>]+)>)/gi;

  const alert_cause = (typeof alert.cause !== 'undefined')
    ? (<>{alert.cause} - </>)
    : (<></>);

  const alert_effect = (typeof alert.effect !== 'undefined')
    ? (<>{alert.effect} - </>)
    : (<></>);

  return (
    <div className="card mb-3">
      <div className="card-header alert-item-header" style={alertStyle}>
        <strong><Link to={`/routes/${route.route_gid}`}>{route.route_short_name} - {route.route_long_name}</Link></strong>
      </div>
      <div className="card-body alert-item-text">
        <p>
          {dayjs().isBefore(dayjs.unix(alert.active_period[0].start)) && (
            <span><span className="badge bg-secondary text-bg-secondary">Future</span>&nbsp;</span>
          )}
          <strong>{alert.header_text.translation[0].text.replace(htmlCleanupRegex, '')}</strong>
        </p>
        {alert.description_text.translation[0].text.replace(htmlCleanupRegex, '')}
        {alert.informed_entity.filter((ie) => typeof ie.stop_id !== 'undefined').length > 0 && (
          <>
            <hr />
              <ul className="list-inline">
                <li className="list-inline-item"><strong>Stops:</strong></li>
                {alert.informed_entity.map((item, key) => (<li key={key} className="list-inline-item"><StopCode stop={{ stop_code: item.stop_id }}/></li>))}
            </ul>
          </>
        )}
      </div>
      <div className="card-footer alert-item-footer" style={alertStyle}>
        {alert_cause}
        {alert_effect}
        <strong>Start:</strong> {renderUnixTimestamp(alert.active_period[0].start)}
        {(alert.active_period[0].end && alert.active_period[0].end < 32503701600)
          && <>&nbsp;-&nbsp;<strong>End:</strong> {renderUnixTimestamp(alert.active_period[0].end)}</>
        }
      </div>
    </div>
  );
}

AlertItem.propTypes = {
  alert: PropTypes.object.isRequired,
  route: PropTypes.object,
};

AlertItem.defaultProps = {
  route: {
    route_color: '#999',
    route_long_name: 'Route Unavailable',
    route_short_name: '00',
  },
};

export default AlertItem;
