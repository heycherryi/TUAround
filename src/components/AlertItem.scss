.alert-item-header {
  a:any-link {
    color: inherit;
    text-decoration: none;
  }
}

.alert-item-text {
  white-space: pre-line;
  font-size: 0.8rem;
}

.alert-item-footer {
  text-align: right;
  font-size: 0.75rem;
}
