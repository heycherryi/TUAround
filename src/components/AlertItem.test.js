/* globals test, expect */

import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import AlertItem from './AlertItem';

const alertsFixture = require('../fixtures/alerts.json');
const routesFixture = require('../fixtures/routes.json');

test('renders AlertItem', () => {
  const route = routesFixture.data.find((i) => i.route_gid === alertsFixture[0].alert.informed_entity[0].route_id);
  const { container } = render(
    <Router>
      <AlertItem alert={alertsFixture[0].alert} route={route} />
    </Router>,
  );
  expect(screen.getByText('79 - SKYLINE')).toBeInTheDocument();
  expect(screen.getByText('4/4/22, 2:24 AM')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test('renders AlertItem with informed stops', () => {
  const route = routesFixture.data.find((i) => i.route_gid === alertsFixture[3].alert.informed_entity[0].route_id);
  const { container } = render(
    <Router>
      <AlertItem alert={alertsFixture[3].alert} route={route} />
    </Router>,
  );
  expect(screen.getByText('14 - WHITES CREEK')).toBeInTheDocument();
  expect(screen.getByText('WHIDICSN')).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});
