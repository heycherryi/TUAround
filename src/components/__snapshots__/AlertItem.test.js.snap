// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders AlertItem 1`] = `
<div>
  <div
    class="card mb-3"
  >
    <div
      class="card-header alert-item-header"
      style="border-color: #5d5f62; background-color: rgb(93, 95, 98); color: rgb(255, 255, 255);"
    >
      <strong>
        <a
          href="/routes/79"
        >
          79
           - 
          SKYLINE
        </a>
      </strong>
    </div>
    <div
      class="card-body alert-item-text"
    >
      <p>
        <strong>
          Detour in effect on route 79 SKYLINE NORTHBOUND
        </strong>
      </p>
      Detour in effect on route 79 SKYLINE

Route 79 NORTHBOUND will detour 

Detour Route 79 from 
Madison to N Dickerson
No service on Skyline Ridge due to construction
From Skyline Hospital, continue Doverside 
back to Dickerson
R-Dickerson to regular route. 
    </div>
    <div
      class="card-footer alert-item-footer"
      style="border-color: #5d5f62; background-color: rgb(93, 95, 98); color: rgb(255, 255, 255);"
    >
      Construction
       - 
      Detour
       - 
      <strong>
        Start:
      </strong>
       
      <span
        title="1649057040"
      >
        4/4/22, 2:24 AM
      </span>
    </div>
  </div>
</div>
`;

exports[`renders AlertItem with informed stops 1`] = `
<div>
  <div
    class="card mb-3"
  >
    <div
      class="card-header alert-item-header"
      style="border-color: #753cbe; background-color: rgb(117, 60, 190); color: rgb(255, 255, 255);"
    >
      <strong>
        <a
          href="/routes/14"
        >
          14
           - 
          WHITES CREEK
        </a>
      </strong>
    </div>
    <div
      class="card-body alert-item-text"
    >
      <p>
        <strong>
          Detour in effect on route 14 WHITES CREEK TO DOWNTOWN
        </strong>
      </p>
      Detour in effect on route 14 WHITES CREEK

Route 14 TO DOWNTOWN will detour 

Inbound Only.
Continue FERN AVE (E)
R-DICKERSON PIKE
Back to regular route.

Stops between BRICK CHURCH PIKE & FERN AVE NB and N DICKERSON PIKE & CLEVELAND ST SB will not be served while this detour is in effect.
      <hr />
      <ul
        class="list-inline"
      >
        <li
          class="list-inline-item"
        >
          <strong>
            Stops:
          </strong>
        </li>
        <li
          class="list-inline-item"
        >
          <span
            class="stop-code badge bg-white text-black border border-secondary"
          >
            <a
              href="/stops/WHIDICSN"
            >
              WHIDICSN
            </a>
          </span>
        </li>
      </ul>
    </div>
    <div
      class="card-footer alert-item-footer"
      style="border-color: #753cbe; background-color: rgb(117, 60, 190); color: rgb(255, 255, 255);"
    >
      Other Cause
       - 
      <strong>
        Start:
      </strong>
       
      <span
        title="1654988700"
      >
        6/11/22, 6:05 PM
      </span>
    </div>
  </div>
</div>
`;
